'FetchBlocks' gateway option

Voker57 opened the pull request:
If FetchBlocks=false, don't fetch blocks which are not already present in storage.

Also add Gateway.FetchBlocks option

This mode can be useful for creating gateways which are supposed to serve only chosen content.

Kubuxu commented:
There is already `ipfs daemon --offline`. Is it not good enough?

Voker57 commented:
@Kubuxu --offline will also disable fetching via API. I want only gateway to be restricted.

magik6k commented:
I think it might be better to implement this as a config option

whyrusleeping commented on cmd/ipfs/daemon.go:
```diff
@@ -148,6 +149,7 @@ Headers.
 		cmds.BoolOption(initOptionKwd, "Initialize ipfs with default settings if not already initialized").Default(false),
 		cmds.StringOption(routingOptionKwd, "Overrides the routing option").Default("dht"),
 		cmds.BoolOption(mountKwd, "Mounts IPFS to the filesystem").Default(false),
+		cmds.BoolOption(offlineGatewayKwd, "Do not fetch blocks because of gateway requests").Default(false),
```
s/because of/for/

magik6k commented:
IMO daemon already has too many flags (`--enable-gc` as one example that could be in config). Gateway already has it's config scope, this could be done in config as `Gateway.Offline`. I'd say that all options that make sense in live deployments should be in config.

Stebalien commented:
How about something like: https://github.com/ipfs/go-ipfs/issues/4180

lgierth commented:
Agreed that a config option is desirable over a command flag.

I'm afraid "offline" might not be the best name here (and you're also inconsistent naming it "InternalDAG". We're already giving "offline" way too many meanings in various contexts.

lgierth commented:
Oh FFS where did my comment go? The form apparantly reset to a very early stage of it...

lgierth commented:
Phew I recovered it -- deleting my two comments above.

-----

Good idea :+1: 

Agreed that a config option is desirable over a command flag, and we'll also want to refactor the gateway a bit. It should get a GatewayOpts struct, similar to the HostOpts struct I implemented in https://github.com/libp2p/go-libp2p/pull/197

I'll put up to debate whether "offline" is a good name though. It already means a couple of different things in go-ipfs. This option seems spiritiually related to the `ipfs add --local` and `ipfs name resolve --local` flags, in that it only operates locally (apart form IPNS lookups I guess). I'm not ready to say one is better of the other, just marking this as an important bit. The reason for it's importance is probably outside of this PR and more in the already existing meanings of "offline". It's important the proposed feature fits in nicely with the existing modes of connectedness and disconnectedness and doesn't unneccessarily complicate them.

lgierth commented:
Btw, mentioning the `--local` flag, I noticed that its description in `ipfs --help` is completely inaccurate. See #4182.

Voker57 commented:
@lgierth Isn't GatewayConfig already doing same job as proposed GatewayOpts? https://github.com/ipfs/go-ipfs/blob/master/core/corehttp/gateway.go#L14

lgierth commented:
Yeah you're right that's already useful

whyrusleeping commented:
Maybe the better term would be `NoFetch` as a field inside the gateway options?

Voker57 commented:
Renamed option to FetchBlocks and moved to config.

nicola commented:
I want this so badly :(

Voker57 commented:
@nicola yes.

Kubuxu commented:
Where is the option handled in the gateway? I see it only added to ServeOptions.

Voker57 commented:
Sorry, I screwed up a rebase. Will update.

Kubuxu commented:
Also:
```
+ ./bin/gx deps dupes

+ test -z package go-ipfs-config imported as both QmSoYrBMibm2T3LupaLuez7LPGnyrJwdRxvTfPUyCp691u and QmQwsBSJXHs3nHBLPJKd1iBNpW9uJ9acSpqbbjafvNJQc3
```

and as tests passed on circleci and travis, they seem to be too weak. I would recommend using IPTB to setup 2 node cluster.

Voker57 commented:
Fixed the PR and improved the tests.

Stebalien commented on core/coreapi/coreapi.go:
```diff
@@ -16,19 +16,26 @@ package coreapi
 // NewCoreAPI creates new instance of IPFS CoreAPI backed by go-ipfs Node.
-func NewCoreAPI(n *core.IpfsNode) coreiface.CoreAPI {
-	api := &CoreAPI{n}
+func NewCoreAPI(n *core.IpfsNode, offlineMode bool) coreiface.CoreAPI {
```
This is really not the right place to configure this.

Voker57 replied:
why not? I don't see how @magik6k solution is better, except for modifying less code.

Stebalien replied:
In retrospect, it's probably a reasonable place to do this (although I would have used functional options). Really, I was reacting to the fact that the CoreAPI needed to store a separate DAGService (I don't want it to turn into yet another "IpfsNode"). However, it turns out that we have to do that for sessions (no way around that as far as I can tell given the current sessions design) so this really isn't an issue.

Stebalien commented on core/coreapi/coreapi.go:
```diff
@@ -16,19 +16,26 @@ package coreapi
 
 type CoreAPI struct {
 	node *core.IpfsNode
+	dag  ipld.DAGService
```
This is going to be a footgun. Now the CoreAPI needs to make sure to use *it's* dag instead of the global one.

Maybe https://github.com/ipfs/go-ipfs/issues/4009? This keeps on haunting us and I've yet to see a better proposal.

magik6k commented:
I'll handle this

Voker57 commented:
Rebased and updated to use new API. It is now similar to #5649 but more limited in scope: will still resolve IPNS fully, which is my desired behavior. Not sure how to cover this by tests, ones included in this PR would still pass but IPNS will not be queried fully with #5649.

magik6k commented on core/coreapi/coreapi.go:
```diff
@@ -217,6 +216,10 @@ func (api *CoreAPI) WithOptions(opts ...options.ApiOption) (coreiface.CoreAPI, e
 	}
 
+	if !settings.FetchBlocks {
+		subApi.dag = dag.NewDAGService(bserv.New(subApi.blockstore, offlinexch.Exchange(subApi.blockstore)))
```
`settings.FetchBlocks` needs to set `exchange`/`blocks` too, otherwise it still will be possible to use gateway to fetch stuff from the network with things like `/api/v0/block/get`

magik6k commented on core/corehttp/gateway.go:
```diff
@@ -8,13 +8,15 @@ import (
 type GatewayConfig struct {
 	Headers      map[string][]string
 	Writable     bool
+	FetchBlocks  bool
```
Is this used anywhere?

Voker57 replied:
no, leftover.

magik6k commented:
IPNS should now work in https://github.com/ipfs/go-ipfs/pull/5649, can you check if it works for you?

Voker57 commented:
> IPNS should now work in #5649, can you check if it works for you?

yes, now this PR works like I need it. I'm closing this one.
